packages/backend-core/src/objectStore/objectStore.ts

```diff
@@ -251,14 +251,35 @@ export async function streamUpload({
       : CONTENT_TYPE_MAP.txt
   }
 
+  const bucket = sanitizeBucket(bucketName),
+    objKey = sanitizeKey(filename)
   const params = {
-    Bucket: sanitizeBucket(bucketName),
-    Key: sanitizeKey(filename),
+    Bucket: bucket,
+    Key: objKey,
     Body: stream,
     ContentType: contentType,
     ...extra,
   }
-  return objectStore.upload(params).promise()
+
+  // make sure we have the stream before we try to push it to object store
+  if (stream.on) {
+    await new Promise((resolve, reject) => {
+      stream.on("finish", resolve)
+      stream.on("error", reject)
+    })
+  }
```

Comment:
Does this not defeat the purpose of streaming? The idea is to never hold more than a small amount of the object in memory. If we're waiting for the stream to end, we're going to be buffering the whole object before uploading it.

Reply:
So I thought this as well - however it seems like when we pass the stream onto S3 you only get a partial of the file, as it simply streams up to where-ever the stream was at - I tried a few ways to get around this, but we need the stream to be complete before we can pass it to S3 to avoid partial uploads of files - I couldn't find a lot of information about this though, it seems to be a very niche problem and I don't know how we've never ran into this before.

Considered making the wait for finish a flagged option so that we only use it when working with web streams (pulled through node-fetch).

Reply:
I don't think we can allow this into production as-is. We have a lot of memory problems already, and they appear to be part of what causes container restarts.

I'd instead opt to stream these files first to disk, and then from disk to S3. This would keep memory use low at the cost of ephemeral disk. The pods in cloud today don't have any ephemeral disk in their resources config, but I can change that for you if we go down this road.

Reply:
Updated to use `tmp` to write the file through disk rather than straight as a stream.